### ISSUE TITLE
IntelliSense includePath recursive

### ISSUE THREAD

**Issue by johmarjac** [microsoft/vscode-cpptools#2141]
Hi,

I have installed VSCode on my Debian 9 VM, cloned the betaflight/betaflight repository from GitHub opened that repository folder in VSCode and saved a Workspace File in that betaflight directory.

I setup tasks.json to call shell command `make TARGET=OMNIBUS`.
When I run that tasks it builds the whole repository just fine. So that is working fine.
Then I installed vscode-cpptools because I wanted to start modifying files with proper IntelliSense.

The problem is, that those include files are in thousand sub-directories and I dont want to put every single directory into the includePath of my c_cpp_properties.json.

So I dont get why it cant be recursively searched for headers?
Literally all the included headers have that green lightbulb. Also on documentation it says you can click on the lightbulb to get suggestions for including, but if I click on it nothing happens, just hovering on it creates a separate small tooltip explaining the error in more detail, but there is nothing to click at all.

Why is it so hard to include headers in VSCode Extension Cpp? This makes me almost go back to using Eclipse for this...

**Comment by bobbrow** [comment]
You need to use `"${workspaceFolder}/**"` for your `"includePath"` to turn on the recursive include feature.  It is currently off by default because we are still optimizing it for larger folders.  We do plan to have a release out today or tomorrow with the performance improvements, but if you want to start trying out the feature, we have some of the optimizations (but not all) released already in our [insiders build here](https://github.com/Microsoft/vscode-cpptools/releases/tag/v0.17.5-insiders)

**Comment by johmarjac** [comment]
Hi @bobbrow thanks for the answer 👍 

I am currently using 0.17.4, is that feature with /** available there? So basically if I do:

`${workspaceFolder}` will include only headers in same directory
`${workspaceFolder}/**` will include all headers in all sub directories right?

Then next thing will be, how can I exlude directories? Does that work? Because
I have a `target` directory which contains headers that are required, but that folder also contains folders for all the different controllerboards supported but only one is required at compile time.

That means

`${workspaceFolder}/src/main/target/` is required to include the required headers in there
and one single folder is required:
`${workspaceFolder}/src/main/target/OMNIBUS/`

but in
`${workspaceFolder}/src/main/` there is dozens of other folders containing necessary includes.

If I put 
`${workspaceFolder}/src/main/**` it will also include all the controller boards in /target folder right?
So I am looking for something like exclude a specific folder.

Thanks in advance

**Comment by bobbrow** [comment]
Yes, the feature is available in 0.17.4, but it is not well optimized for large folders, and I don't think exclusions support (via VS Code's `files.exclude` setting) was turned on yet either.  Both are addressed in 0.17.5.  If you prefer a particular path over the recursive one, just add it before the recursive path and it will be preferred over any of the other boards:

```
    "includePath": [
        "${workspaceFolder}/src/main/target/OMNIBUS/",
        "${workspaceFolder}/**"
    ]
```

In some cases this might be easier than setting up a ton of excludes.

**Comment by sean-mcmanus** [comment]
I believe 0.17.4 supports files.exclude.
